### DISCUSSION TITLE
Hosted exit status conditions?

### DISCUSSION THREAD

**Discussion by SirWumpus** [ForthHub/discussion#185] [General]
For a hosted environment, should the standard document non-zero exit status conditions?

**Comment by ruv** [comment]
There are many Forth systems running on operating systems and providing the similar capabilities in this regard. It is good for the entire ecosystem if such capabilities are available through the **same interface**. For example, same sub-set of common command line options, same sub-set (with the same meaning) of common [exit statuses](https://en.wikipedia.org/wiki/Exit_status).

My attempt: [Desktop Forth system requirements](https://github.com/ForthHub/discussion/wiki/Desktop-Forth-system-requirements) (in the ForthHub wiki). See also the corresponding [discussion](https://github.com/ForthHub/discussion/issues/83).

I think, a proposal can be development for that.

**Reply by SirWumpus** [reply]
Hmm.  Well before a proposal maybe start a discussion so I'll kick off with a minimum set:
```
EXIT STATUS
0       Forth system and application exited successfully.
1       The application returned a negative result.
2       Forth system or application command line usage error.
3       One or more exceptions occurred from either the Forth system or application.
>3      Some Forth application error occurred.
```
All the SUS conforming shells allow for errors 1..125 and reserve the remaining for the shell, see [SUS sh(1)](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/sh.html#tag_20_117_14)

Unlike the Forth Standard's predilection for adding an exception for everything (like -255 kitchen sink inaccessible), exit status codes should be kept to a minimum as most usage cases just care if the Forth system or application succeeded or failed.

- - -
> My attempt: [Desktop Forth system requirements](https://github.com/ForthHub/discussion/wiki/Desktop-Forth-system-requirements) (in the ForthHub wiki)

Your recommended minimum options assume a GNU CLI option parser WRT `--help` and `--version`.  The POSIX `getopt()` does not support double-hyphen options.  I believe a hosted Forth should conform to the SUS (POSIX.2) [Utility Syntax Guidelines](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html#tag_12_02).

Typically unknown options display a short usage summary message to standard error.  The character `?` intentionally cannot be used for an option so `-?` is the equivalent of GNU's `--help`.

The `-v` is popular for a verbose or debug output option, so instead a version option should be `-V`.

For GNU fans they can support long option names `--help` and `--version` as alternatives to the short name.

**Reply by ruv** [reply]
> EXIT STATUS
> 0       Forth system and application exited successfully.
> 1       The application returned a negative result.

If we talk about standardizing, the notion "application" is not suitable. In addition, the exit status for a Forth program itself is outside the scope of the standard.

I can suggest the following:

```
EXIT STATUS
0       Normal termination.
1       Unspecified abnormal termination.
2       Incorrect command line argument.
3       Uncaught exception occurs.
```

"Normal termination" can occur in the following conditions:
- `BYE` is called,
- the end of the input stream is reached in `QUIT`,
- no more command for processing (in the batch mode).

"Abnormal termination" can occur if the program has entered in an _ambiguous condition_, and the Forth system [has taken](https://forth-standard.org/standard/usage#usage:ambiguous) to terminate.

"Incorrect command line argument" — a command line argument is not recognized.

"Uncaught exception" — an exception occurs that the program does not catch while the system runs in the batch mode.

**Reply by SirWumpus** [reply]
I agree with the bulk of the rewording.

> "Uncaught exception" — an exception occurs that the program does not catch while the system runs in the batch mode.

My only issue here is that it should NOT be limited to batch mode.  Interactive should return same for consistency (and distinguishing between batch and interactive might be problematic in some environment); its probably not an issue since most users will not care about the exit status when interactive.

**Reply by ruv** [reply]
I think,  termination with `abort` can have own exit status. Because `abort` simply tells the system that the program doesn't want to continue execution.

```
EXIT STATUS
4       Program aborts
```

"Program aborts" — an uncaught exception with code `-1` or `-2` occurs in the batch mode.

-----

The associations of exit statuses and meanings can be changed.

**Reply by SirWumpus** [reply]
> I think, termination with `abort` can have own exit status. Because `abort` simply tells the system that the program doesn't want to continue execution.

Disagree.  Having a separate exit status provides little aid IMO when scripting.  `ABORT` is still an exception and is easily covered by exit 3.  If a batch job wants to return something more specific, then catching the exception and terminating with something like `-1 = IF 4 BYE-STATUS THEN`.

**Reply by ruv** [reply]
> My only issue here is that it should NOT be limited to batch mode. Interactive should return same for consistency (and distinguishing between batch and interactive might be problematic in some environment)

Interactive mode means that the system runs `QUIT` (or its equivalent) after processing the command line options.  In this mode, if an uncaught exception occurs, it is handled by the system as specified in [9.6.1.2275 `THROW`](https://forth-standard.org/standard/exception/THROW). And there is no place to terminate except with `bye` (or its equivalent), end of the input stream, or something like `Ctrl+C`, or due to an ambiguous condition.

Termination by Ctrl+C (if any) should have the exit status 4 "Abort", or maybe its own exit status.

-----

Concerning Ctrl+C, the best behavior is following: if it is pressed when the system waits for input in `QUIT`, then the system terminates. Otherwise the system performs `QUIT`.

**Reply by ruv** [reply]
>  `ABORT` is still an exception and is easily covered by exit 3. If a batch job wants to return something more specific, then catching the exception and terminating with something like `-1 = IF 4 BYE-STATUS THEN`.

`ABORT` for a program is similar to `BYE` for the process.   Both are used for **intended** premature termination of execution.

Typical usage:

```forth
: main
  "some-external-feature" feature-available ( flag )
  0= abort" some-external-feature is not available"
  ... \ use the feature
;
```

Or like this:
```forth
[defined] cool-word [if]
  run-program
[else]
  .( `cool-word` is required, Abort. ) abort
[then]
```

So it is not a condition of a program error. But an uncaught exception means that there an exception occurs that is **not handled** properly by the program.

I think, it is better to distinguish these cases.

It is also similar to Ctrl+C by the user/operator.    So, I think, `abort` and `Ctrl+C` should both have the same exist status.

**Reply by SirWumpus** [reply]
> Termination by Ctrl+C (if any) should have the exit status 4 "Abort", or maybe its own exit status.

Disagree.

> Concerning Ctrl+C, the best behavior is following: if it is pressed when the system waits for input in `QUIT`, then the system terminates. Otherwise the system performs `QUIT`.

I disagree with the idea of having split behaviour dependent on input state.  CTRL+C (SIGINT) should behave consistently; CTRL+C should result in `-28 User Interburp` being thrown and following the processing of `ABORT` as specified, not `QUIT`.  

If you want to force termination there is SIGQUIT CTRL+\ or SIGTSTP CTRL+Z plus `kill %1` provided by the OS (assuming those signals are not trapped).  (Sure I'm thinking in terms of POSIX systems, because its well understood and parallels for other environments I'm sure can be drawn, though Windows is always painful.)

**Reply by ruv** [reply]
Concerning `BYE-STATUS` — this will terminate the process. But the idea of `abort` is that the system takes control back. In the interactive mode the user might want to check something interactively (after `abort`).  In the batch mode the system should terminate.

**Reply by SirWumpus** [reply]
In an interactive system (not necessarily Forth) where CTRL+C is handled, I expect it to terminate processing and return to the REPL prompt, then I can terminate by EOF (CTRL+D), `exit`, `quit`, `get out of the house`, or what ever command terminates the process.

ABORT terminates processing and returns to the REPL prompt.  Even in a batch job it behaves that way as I see it.  It just commonly happens that when an explicit `ABORT` occurs there is typically no further input so the batch process also terminates.

**Reply by SirWumpus** [reply]
> Concerning `BYE-STATUS` — this will terminate the process. But the idea of `abort` is that the system takes control back. In the interactive mode the user might want to check something interactively (after `abort`). In the batch mode the system should terminate.

In batch mode I might want to stop processing with `ABORT` and drop into interactive mode to do some debugging too (especially for long jobs).  Or I might want to `ABORT` and then also script commands after to automate the debugging if I need to repeat testing.

I think for batch mode, explicit `BYE`, `BYE-STATUS`, EOF, or external signal should terminate the process.  Everything else is scriptable and the process continues to run.

**Reply by ruv** [reply]
> In an interactive system (not necessarily Forth) where CTRL+C is handled, I expect it to terminate processing and return to the REPL prompt,

Some systems terminates by Ctrl+C.  For example, mysql-client, node (by double Ctrl+C). Some — does not (e.g., [GHCi](https://downloads.haskell.org/ghc/latest/docs/users_guide/intro.html), python3).  Some Forth systems terminate on Ctrl+C, some — does not.

I don't have a strong opinion on this.  Probably, termination on Ctrl+C makes the system more comfortable for newbies.


> ABORT terminates processing and returns to the REPL prompt. 

Yes, if REPL available.  I would prefer to have a nonzero exit status if `abort` occurs in the batch mode.

**Reply by SirWumpus** [reply]
> > In an interactive system (not necessarily Forth) where CTRL+C is handled, I expect it to terminate processing and return to the REPL prompt,
> 
> Some systems terminates by Ctrl+C. For example, mysql-client, node (by double Ctrl+C). Some — does not (e.g., [GHCi](https://downloads.haskell.org/ghc/latest/docs/users_guide/intro.html), python3). Some Forth systems terminate on Ctrl+C, some — does not.
> 
> I don't have a strong opinion on this. Probably, termination on Ctrl+C makes the system more comfortable for newbies.
> 

Well in my mind I was actually thinking of old BASIC interpreters (but didn't want to embarrass myself).

> > ABORT terminates processing and returns to the REPL prompt.
> 
> Yes, if REPL available. I would prefer to have a nonzero exit status if `abort` occurs in the batch mode.

And a single exit 3 for ANY exception works just nicely.  Having a separate code for `ABORT` brings what?  If I'm shell scripting with Forth doing some intermediate processing I'm most likely going to say:

```shell
if ! $(printf ".( do some stuff ) ' not-a-word .( oops )" | post4) ; then 
    echo oopsie
fi

if ! $(printf ".( do some stuff ) abort .( debug here )" | post4) ; then 
    echo "I have a cunning plan!"
fi
```
Do I really care what type of exception?

**Reply by ruv** [reply]
> Do I really care what type of exception?

Sometimes it makes sense to assess how critical an error is. Otherwise only two values would be enough: 0 and 1.

The conditions of terminating a Forth system can be ordered from less to more critical (more severe) as follows:

- Normal termination
- Abort
- Unrecognized command line option
- Error while compiling a definition
- Uncaught exception
- Unspecified abnormal termination

If we want to use fewer different exit statuses, we can group multiple conditions under one exit status.

**Reply by ruv** [reply]
A possible approach is that for positive statuses, some lower status numbers cover the cases of all higher status number.

For example:
- `0` — Normal termination
- `1` — Unspecified abnormal termination
- `2` — Uncaught exception
- `3` — Static error in program
- `4` — Unrecognized command line argument
- `5` — Abort

[Throw codes](https://forth-standard.org/standard/exception#table:throw) can be linked as follows:
- Static error in program: `-13`, `-14`, `-16`, `-18`, `-19`, `-22`, `-27`, `-29`, `-30`, `-32`, `-39`, `-48`, `-51`
- Abort: `-1`, `-2`, `-28`, `-56`

If the Forth system **can** detect conditions for multiple statuses, it should use the exit status with the highest number among them.

Which conditions a Forth system can detect/distinguish is **implementation defined**.

**Reply by ruv** [reply]
Perhaps, a better order (in terms of detecting different conditions, and "nesting" of them):
- 0 — Normal termination
- 1 — Unspecified abnormal termination
- 2 — Uncaught exception
- 3 — Abort
- 4 — Static error in program
- 5 — Unrecognized command line argument

**Reply by SirWumpus** [reply]
I think of grep(1) often when I think of exit status:

* 0 Success; match found
* 1 No match found, no answer, "I'm sorry Dave, I'm afraid I can't do that"
* 2 Command line usage error, eg. invalid option, bad regex.; program can't start
* 3 Uncaught exceptions including Abort
* \>3 implementation or user program errors

I prefer exit 1 to reflect an ***intended*** negative result.  Exit 2 for usage errors (bad command line options, bad regex, smells like fish, missing ingredients, things needed to start/initialise) since that is a frequent category of error.

Exit 3 : I see no need to need to distinguish between Uncaught Exception and Abort; an Abort is an Uncaught Exception.  I expect Uncaught Exceptions to be the next most common error during the life of the process or user program. 

Exit >3 should be either implementation defined and/or for the user program. 

Slicing exits status into finer granularity seems pointless.  The user can either view the console (or logs) for detailed error messages .

**Reply by ruv** [reply]
My idea is that if the Forth system does not distinguish different conditions, it can just return the exit status `1`.  That is, this status covers all cases of abnormal termination.

In general, less specific conditions cover (include) more specific conditions. Thus, if the system does not distinguish a more specific case, it uses the exit status for a more general case (with a lower exit status number).

The case "Unspecified abnormal termination" includes the case of "Uncaught exception" and cases of ambiguous conditions when the system's choice is to terminate.

The case  "Uncaught exception" includes any exception, and also the cases of "Abort", "Error in program", "Error in command line arguments", etc.

The case "Static error in program" includes the case "Unrecognized command line argument".

So, it's up to the system what conditions it will distinguish.

Exit status numbers {16, ..., 31} could be reserved for system-defined conditions.

When a program is loaded, it is possible that the system is not a standard system (see [4.2.2](https://forth-standard.org/standard/doc#subsection.4.2.2)). In this case, the meanings of exit status numbers are up to the program.

**Comment by vygr** [comment]
Can Chuck explain how I leave this club ? Or is it one of those once your in, there’s no going out deals ?ChrisPs. You Forth folks should learn Lisp. And then ChrysaLisp and why it doesn't do what Lisp folks say to do ;) prebinder builds a dictionary…..—————-Chris Hinsley On 5 Nov 2024, at 11:35 AM, Anthony Howe ***@***.***> wrote:﻿
Hmm.  Well before a proposal maybe start a discussion so I'll kick off with a minimum set:
EXIT STATUS
0       Forth system and application exited successfully.
1       The application returned a negative result.
2       Forth system command line usage error.
3       One or more exceptions occurred from either the Forth system or application.
>3      Some Forth application error occurred.

All the SUS conforming shells allow for errors 1..125 and reserve the remaining for the shell, see SUS sh(1)
Unlike the Forth Standard's predilection for adding an exception for everything (like -255 kitchen sink inaccessible), exit status codes should be kept to a minimum as most usage cases just care if the Forth system or application succeeded or failed.


My attempt: Desktop Forth system requirements (in the ForthHub wiki)

Your recommended minimum options assume a GNU CLI option parser WRT --help and --version.  The POSIX getopt() does not support double-hyphen options.  I believe a hosted Forth should conform to the SUS (POSIX.2) Utility Syntax Guidelines.
Typically unknown options display a short usage summary message to standard error.  The character ? intentionally cannot be used for an option so -? is the equivalent of GNU's --help.
The -v is popular for a verbose or debug output option, so instead a version option should be -V.
For GNU fans they can support long option names --help and --version as alternatives to the short name.

**Reply by ruv** [reply]
> Can Chuck explain how I leave this club ?

1. Sign In to your GitHub account and open Settings/Organizations at https://github.com/settings/organizations
2. Next to "ForthHub" click **Leave**.

See also the GitHub documentation [Removing yourself from an organization](https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-your-membership-in-organizations/removing-yourself-from-an-organization).

**Comment by vygr** [comment]
So I have to leave the organisation in order to stop the spam ?That’s not very fair. I don’t want to leave the organisation. I just want to stop listening to your whineings. Chris—————-Chris Hinsley On 5 Nov 2024, at 4:00 PM, ruv ***@***.***> wrote:﻿

Can Chuck explain how I leave this club ?


Sign In to your GitHub account and open Settings/Organizations at https://github.com/settings/organizations
Next to "ForthHub" click Leave.

See also the GitHub documentation Removing yourself from an organization.

**Reply by ruv** [reply]
> I don’t want to leave the organisation. I just want to stop listening to your whineings.
 
@vygr, your messages are off-topic.  You can **Unwatch** the repository [ForthHub/discussion](https://github.com/ForthHub/discussion). Otherwise, use Google to find out how to unsubscribe from something you've subscribed to on GitHub.

**Comment by niclash** [comment]
If you go to https://github.com/watching, then you can add repositories 
to not be notified for. If you are not a GitHUb user for other things, 
you have more broad Notification options in 
https://github.com/settings/notifications

HTH
Niclas

**Comment by ruv** [comment]
A citation from [Post4 / Issue#54](https://github.com/SirWumpus/post4/issues/54#issuecomment-2457117548):
> For example piping input `echo ABORT | post4`; I want to see a non-zero exit status because there was an exception at some point during the session. So `echo no-such-word | post4` will result in -13 exception and a non-zero exit status.

```shell
echo " ' abort catch " | post4
echo -e " abort \n bye" | post4
```
Do you count above cases as "there was an exception at some point"?

If you mean only counting uncaught exceptions in interactive mode, then it's still a moot point.  It will require adding a variable and counting these exceptions.   And what the use?

In production (if not testing), multi-line scripts should not be run in interactive mode (i.e., be passed as the user input device).

**Reply by SirWumpus** [reply]
> A citation from [Post4 / Issue#54](https://github.com/SirWumpus/post4/issues/54#issuecomment-2457117548):
> 
> > For example piping input `echo ABORT | post4`; I want to see a non-zero exit status because there was an exception at some point during the session. So `echo no-such-word | post4` will result in -13 exception and a non-zero exit status.
> 
> ```shell
> echo " ' abort catch " | post4
> echo -e " abort \n bye" | post4
> ```
> 
> Do you count above cases as "there was an exception at some point"?
> 
> If you mean only counting uncaught exceptions in interactive mode, then it's still a moot point. It will require adding a variable and counting these exceptions. And what the use?

I mean uncaught exceptions.  And by counting I mean more like set a flag, not actually count 1, 2, 3 ...

> In production (if not testing), multi-line scripts should not be run in interactive mode (i.e., be passed as the user input device).

I disagree.  I could 
```
elf$ post4 script.p4
elf$ post4 < script.p4
elf$ post4 <<EOF
.( Say something for
the folks watching
at home ) 
EOF
elf$
```
equally.

**Reply by ruv** [reply]
>> In production (if not testing), multi-line scripts should not be run in interactive mode (i.e., be passed as the user input device).
>
> I disagree. 

In interactive mode, when an error occurs, the Forth system eventually discards the parse area (the remaining part of the input buffer) and reads the new line from the input stream (that is the user input device).

So if a script file is used as the user input device (by redirection), and an error occurs in the middle of the script (for example, a word is not found, or `r/o open-file throw` and the file is absent),  then interpreting of the remaining part of the script will probably produce more errors.  This does not make any sense in production.

Just an illustration of what I mean:
```shell
post4 <<EOF
: foo
  -#123 .
;
foo
EOF
```
There is no point in continuing to interpret the input file after the first error.  Therefore, you probably will not pass a script to the Forth system stdin in a production environment.

This also means that from the user's perspective it is better if `post4 script.p4` is equivalent to `post4 -i script.p4` rather than `post4 < script.p4`.

**Reply by SirWumpus** [reply]
> So if a script file is used as the user input device (by redirection), and an error occurs in the middle of the script (for example, a word is not found, or r/o open-file throw and the file is absent), then interpreting of the remaining part of the script will probably produce more errors. This does not make any sense in production.

So a failure from redirected input should terminate at the first uncaught exception, ignoring all further input?

**Reply by ruv** [reply]
> So a failure from redirected input should terminate at the first uncaught exception, ignoring all further input?

Not quite.  I distinguish interactive mode and batch mode.

**In interactive mode**, an uncaught exception (regardless of its source) eventually causes the next line to be read and interpreted from the user input device. **In batch mode**, an uncaught exception (regardless of its source) must cause termination.

It is useful if batch/interactive mode can be selected with a command line argument.  Otherwise, the system should automatically select the mode.

I think, the command line arguments should be evaluated in batch mode by default. It means, any error/exception while evaluating command line arguments should cause termination (if the interactive mode was not selected before that).

Redirection of `stdin`/`stdout` should not affect the mode. So, by default, `stdin` is interpreted in interactive mode.  And if batch mode was selected by a command line argument, then `stdin` is interpreted in batch mode (and a prompt should not be shown, the same as when a file is included).

**Reply by SirWumpus** [reply]
> It is useful if batch/interactive mode can be selected with a command line argument. Otherwise, the system should automatically select the mode.

I do not  think a command line option is necessary, since it should be detectable when stdin is redirected, eg. `isatty(STDIN_FILENO)`.

**Reply by ruv** [reply]
>  it should be detectable when stdin is redirected, eg. `isatty(STDIN_FILENO)`

Yes. But for automated testing (for example, for testing `quit`) it is very useful if you can redirect `stdin` and the system will still be in interactive mode.

**Reply by ruv** [reply]
@SirWumpus [wrote](https://github.com/ForthHub/discussion/discussions/185#discussioncomment-11167365):
> In batch mode I might want to stop processing with `ABORT` and drop into interactive mode to do some debugging too (especially for long jobs).

If a Forth system is started in batch mode, the Forth program cannot drop into interactive mode. Batch mode implies that there is no access to the user's interactive input device at all. For example, have a look at Linux daemons or Windows services.

**Reply by SirWumpus** [reply]
Umm.  I think there is a terminology issue here.  "Batch" mode or "scripted" is nothing like a daemon or service or embedded code, which are very different beasts in my mind.  When I think daemons services, I think servers like HTTP, SMTP, DNS, DHCP, etc.  A batch might be `at` and `cron` jobs or some user script.

**Reply by ruv** [reply]
Well, a cron job is also a good example of a situation where the user's _interactive_ input device is not available. If redirection is available, it's up the the user should the redirected `stdin` be handled in interactive mode or in batch mode.
